### PR TITLE
fix: apply reviews changes of the repository

### PR DESCRIPTION
**Pull request by lriberon**
Add recommendations from the INT-3407 ticket.

What have been done:
- Add godoc for every functions and constants
- Update the versions for the `build` and `release` CI jobs
- Use goreleaser v2
- Makefile `OS_ARCH` is now dynamic (based on your machine)
- Remove usage of `ioutil` package
- Handle body close properly in deferred function
- Create HTTP requests with the Terraform's context
- Split the return for `err` and the `StatusCode != 200`
- Add `ValidateDiagFunc` for `response` and `priority` fields
- Define Timeouts for the CRUD operations on custom rules
- Handle properly the `d.Set(...)` in read operation
- Handle error for `Atoi` conversion
- Bump Golang version (1.16 -> 1.22)

### CODE REVIEW

File: datadome-client-go/client.go [hunk 1]
```diff
@@ -3,25 +3,27 @@ package datadome
...
+	"io"
 	"log"
 	"net/http"
 	"time"
 )
 
-// Default datadome dashboard URL
-const HostURL string = "https://customer-api.datadome.co/1.0/protection"
+// HOST_URL default datadome dashboard URL
+const HOST_URL string = "https://customer-api.datadome.co/1.0/protection"
```
Comment (MickaelDatadome) [inline datadome-client-go/client.go:13]:
```suggestion
const HOST_URL string = "https://customer-api.datadome.co/1.1/protection"
```

File: datadome/resource_custom_rule.go [hunk 2]
```diff
@@ -2,42 +2,72 @@ package datadome
...
 				Required: true,
 			},
-			"response": &schema.Schema{
+			"response": {
 				Type:     schema.TypeString,
 				Required: true,
+				ValidateDiagFunc: func(v any, p cty.Path) diag.Diagnostics {
+					var diags diag.Diagnostics
+					value := v.(string)
+					if !(value == "whitelist" || value == "captcha" || value == "block") {
```
Comment (MickaelDatadome) [inline datadome/resource_custom_rule.go:37]: From the [public doc](https://docs.datadome.co/reference/post_1-1-protection-custom-rules), it is
- allow
- captcha
- block
--> this should be updated

File: datadome/resource_custom_rule.go [hunk 3]
```diff
@@ -2,42 +2,72 @@ package datadome
...
+							Detail:   fmt.Sprintf("%q is not an acceptable priority", value),
+						}
+						diags = append(diags, diag)
+					}
+					return diags
+				},
+				Default: "high",
 			},
-			"endpoint_type": &schema.Schema{
+			"endpoint_type": {
```
Comment (MickaelDatadome) [inline datadome/resource_custom_rule.go:66]: This seems to be a restricted list too:
- web
- api-app-mobile
- rss
- api
- login
- submit

**Review by MickaelDatadome — CHANGES_REQUESTED** [review]
Good improvements !
a few clarifications / changes - are required for me

File: datadome/resource_custom_rule.go (hunk 2 above)
Comment (lriberon) [reply datadome/resource_custom_rule.go:37]: I agree that on the API Reference, it is written "allow".

But if you create your custom rule from Terraform, you will see that it is set as "whitelist" anyway (and can be disturbing if you update your custom rule and don't know this behaviour).
It seems like that is only a "front fix" to respect the audit.

We have 2 options:
- keep "whitelist" in order to don't have changes if you run a `terraform plan`  
- ask Dashboard team to also update this to "allow" in the back

File: datadome/resource_custom_rule.go (hunk 3 above)
Comment (lriberon) [reply datadome/resource_custom_rule.go:66]: Changes applied in https://github.com/DataDome/terraform-provider-datadome/pull/24/commits/966615d2312031016acf1cdb66d9115b5a4c33c3

Add the following:
- `ValidateDiagFunc` to restrict to the enum you mentioned
- `Default` value to `web` as it is specified in the API Specification

File: datadome/resource_custom_rule.go (hunk 2 above)
Comment (lriberon) [reply datadome/resource_custom_rule.go]: Code and example.tf file updated

File: .github/workflows/lint.yml
Comment (MickaelDatadome) [inline .github/workflows/lint.yml:1]: eof missing

**Review by MickaelDatadome — APPROVED** [review]
LGTM !

File: .github/workflows/lint.yml
Comment (lriberon) [reply .github/workflows/lint.yml:1]: I will override the whole CI in the #25, should I do the update now?

**Review by Lauredg — APPROVED** [review]
The changes in this PR are relevant and the explanations are very clear. Thank you!

**Review by MickaelDatadome — COMMENTED** [review]
Not tested, but LGTM !
Thanks for the explanations

**Pull request merged**